forte/data/extractor/extractor.py

```diff
@@ -0,0 +1,230 @@
+    # so that vocab is not directly exposed to
+    # outside user.
+    def size(self):
+        return self.__vocab.size()
+
+    def contains(self, entry: Any):
+        return self.__vocab.contians(entry)
+
+    def items(self):
+        return self.__vocab.items()
+
+    def add_entry(self, entry):
+        self.__vocab.add_entry(entry)
+
+    def entry2id(self, entry):
+        return self.__vocab.entry2id(entry)
+
+    def id2entry(self, idx):
+        return self.__vocab.id2entry(idx)
+
+    def get_pad_id(self):
+        return self.__vocab.get_pad_id()
+
+    def update_vocab(self, pack: DataPack, instance: EntryType):
+        raise NotImplementedError()
+
+    def extract(self, pack: DataPack, 
+            instance: EntryType) -> Feature:
+        raise NotImplementedError()
+
+    def add_to_pack(self, pack: DataPack, instance: EntryType, feature: Feature):
+        raise NotImplementedError()
+
+
+class AttributeExtractor(BaseExtractor):
+    def __init__(self, config: Dict):
+        super().__init__(config)
+        self.attribute = config["attribute"]
+
+    def update_vocab(self, pack: DataPack, instance: EntryType):
+        for entry in pack.get(self.entry, instance):
+            self.add_entry(getattr(entry, self.attribute))
+
+    def extract(self, pack: DataPack, instance: EntryType):
+        data = []
+        for entry in pack.get(self.entry, instance):
+            idx = self.entry2id(getattr(entry, self.attribute))
+            data.append(idx)
+        return Feature(data, self.get_pad_id(), 1)
+
+
+class TextExtractor(AttributeExtractor):
+    def __init__(self, config: Dict):
+        config["attribute"] = "text"
+        super().__init__(config)
+
+
+class CharExtractor(BaseExtractor):
+    def __init__(self, config: Dict):
+        self.max_char_length = getattr(config, "max_char_length", None)
+        super().__init__(config)
+
+    def update_vocab(self, pack: DataPack, instance: EntryType):
+        for word in pack.get(self.entry, instance):
+            for char in word.text:
+                self.add_entry(char)
+
+    def extract(self, pack: DataPack, instance: EntryType):
+        data = []
+        max_char_length = -1
+
+        for word in pack.get(self.entry, instance):
+            tmp = []
+            for char in word.text:
+                tmp.append(self.entry2id(char))
+            data.append(tmp)
+            max_char_length = max(max_char_length, len(tmp))
+
+        if self.max_char_length is not None:
+            max_char_length = min(self.max_char_length, max_char_length)
+
+        return Feature(data, self.get_pad_id(), 2)
+
+
+class AnnotationSeqExtractor(BaseExtractor):
+    def __init__(self, config: Dict):
+        super().__init__(config)
+        self.entry = config["entry"]
+        self.attribute = config["attribute"]
+        self.strategy = config["strategy"]
+        self.based_on = config["based_on"]
+
+    @classmethod
+    def bio_variance(cls, tag):
```

Comment:
If this extractor can only deal with BIO then you probably should open a sub-class for this (or rename this extractor)

Reply:
Will do.